### ISSUE TITLE
Optimise reply by tweet Feature

### ISSUE THREAD

**Issue by namangupta01** [publiclab/plots2#4962]

**Comment by siaw23** [comment]
Any errors? Or pointers where I can have a look?

**Comment by jywarren** [comment]
Yes, a few places - optimization is detailed here: https://github.com/publiclab/plots2/pull/3175#discussion_r263211966

but as to functionality I'm not sure, I do believe we saw a Sentry error for it at one point -- here:

* https://sentry.io/share/issue/cd87dd7040b945eca26c906f3f963a5a/
* https://sentry.io/share/issue/e87f95381f20461e835efea1941eabe1/

```
HTTP::ConnectionError
failed to connect: Connection reset by peer - SSL_connect

      parent_tweet = Client.status(in_reply_to_tweet_id, tweet_mode: "extended")
```

That's on this line: https://github.com/publiclab/plots2/blob/master/app/models/comment.rb#L387

**Comment by jywarren** [comment]
https://sentry.io/share/issue/16c09d47269240bdaa4dd6a100e0c258/ also shows a related error!

**Comment by jywarren** [comment]
Added UserTag query optimization here:

https://github.com/publiclab/plots2/commit/a7ca3e9250e8b6b5a8317149f3658685462ac134

**Comment by namangupta01** [comment]
@jywarren all the sentry issues are related to the outbound requests i.e the request that is going out from the server to fetch the data and our system is not able to get the response.

**Comment by namangupta01** [comment]
Can you try for the query ` Client.search(ENV["TWEET_SEARCH"])` in rails console and check if that it is getting some response or not. If it not then there will be something to be done with server configuration.

**Comment by jywarren** [comment]
Yes, it is -
https://gist.github.com/jywarren/ed5ce5c2256ac35487715223c1a46452

Hmm, what log should we look in?

**Comment by namangupta01** [comment]
Strange, It is working in rails console and the same line of code not working in production.

**Comment by namangupta01** [comment]
I guess this should not happen. @jywarren can you please look into the sentry dashboard and tell the total number of time above event occurs and the last occurrence of these events?

**Comment by jywarren** [comment]
Well, why don't I make you a Sentry account? It looks like there are a few inter-related errors, like this one: https://sentry.io/share/issue/0e925d1fc2ab4cc7b5fca0cce5033149/

Try searching for "twitter" in the issues list there!

**Comment by namangupta01** [comment]
Okay so I searched there and there is only one such event that had occurred. I think we can try this one more time and check if it will again give error or not. What do you think?

**Comment by namangupta01** [comment]
I just replied, let see. Fingers Crossed!

**Comment by namangupta01** [comment]
It neither get added nor any sentry event comes. Strange :/

**Comment by namangupta01** [comment]
So we are getting the replies, just checked in logs.
![Screen Shot 2019-03-21 at 2 14 24 AM](https://user-images.githubusercontent.com/20709985/54717960-6d686a00-4b7f-11e9-9cc5-59a6eeda807c.png)

**Comment by namangupta01** [comment]
Since there are no errors occurs -- after seeing the log file. So there is some bug in the code which is bypassing the comment from being added. And the bug is in `check_and_add_tweets` method of the comment.rb file. There are multiple places where we can check. But first of all, we can check if it is able to get the username from the UserTag table from OAuth login from Twitter. Another way we can do it is with the various debugging statement in the code. What do you think Jeff?

**Comment by jywarren** [comment]
Aha, i found an error in Sentry that could help us? Want to take a look?

**Comment by namangupta01** [comment]
Yeah, why not. Are you talking about https://sentry.io/organizations/publiclab/issues/976905758/?project=1410626&referrer=github_integration ?

**Comment by namangupta01** [comment]
Till the point of retrieving tweets, code is working fine, i just had a look at http://publiclab.org/cron_log.log

**Comment by jywarren** [comment]
Yes, the sentry error shows line 427 of this file:

https://github.com/publiclab/plots2/blob/01c7ace23c5a408c1ab07d082174f694bd5cd52d/app/models/comment.rb#L420-L427

So i wonder if it's failing to get this URL mentioned in Sentry: `SocketError: Failed to open TCP connection to t.co:443 (getaddrinfo: No address associated with hostname)`

That's the shortcode for Twitter. 

https://github.com/publiclab/plots2/blob/01c7ace23c5a408c1ab07d082174f694bd5cd52d/app/models/comment.rb#L397-L399

Yeah, so for example it sees a URL like these:

> @PublicLab @gauravano Thank you to @outreachy @gsoc @RailsGirlsSoC #opensourcediversity and the whole movement for… https://t.co/YUnq7YTOaF

Is it converting our carefully prepared URL with the NID in it into a t.co URL? Actually i haven't found many tweets from us that have that `/n/00000` style URL and ALSO a reply...

https://twitter.com/PublicLab/status/1107717566582214662 but I don't see it in the cron_log?

**Comment by jywarren** [comment]
https://twitter.com/PublicLab/status/1105873695782957057

**Comment by jywarren** [comment]
Sample from the log around this tweet: https://twitter.com/PublicLab/status/1102611975949950981

```

@PublicLab @Mapping4Change This was such a great event! Lots to think about in regards to community air sensing projects.
@PublicLab Hey! We love our DIY microscopes. @laurimnelson ordered another one and you sent a Balloon Mapping Kit o… https://t.co/uyTyvR9PgN
@PublicLab @gsoc Hooray! Another great year of Summer of Code!!! And also we are participating in @outreachy this y… https://t.co/A8xgPYUvzd
RT @jywarren: @PublicLab @gsoc Hooray! Another great year of Summer of Code!!! And also we are participating in @outreachy this year as wel…
RT @jywarren: @PublicLab @gsoc Hooray! Another great year of Summer of Code!!! And also we are participating in @outreachy this year as wel…
@PublicLab @gsoc Wow! Hoping to have a great Summer of Code this year too!
@PublicLab @KyleJwilke The right thing was in the wrong box! This microscope is wonderful. Student built. Great mag… https://t.co/8vRZGBu0zK
RT @laurimnelson: @PublicLab @KyleJwilke The right thing was in the wrong box! This microscope is wonderful. Student built. Great magnifica…
@PublicLab @Mapping4Change This was such a great event! Lots to think about in regards to community air sensing projects.
@PublicLab Hey! We love our DIY microscopes. @laurimnelson ordered another one and you sent a Balloon Mapping Kit o… https://t.co/uyTyvR9PgN
@PublicLab @gsoc Hooray! Another great year of Summer of Code!!! And also we are participating in @outreachy this y… https://t.co/A8xgPYUvzd
RT @jywarren: @PublicLab @gsoc Hooray! Another great year of Summer of Code!!! And also we are participating in @outreachy this year as wel…
RT @jywarren: @PublicLab @gsoc Hooray! Another great year of Summer of Code!!! And also we are participating in @outreachy this year as wel…
@PublicLab @gsoc Wow! Hoping to have a great Summer of Code this year too!
@PublicLab @KyleJwilke The right thing was in the wrong box! This microscope is wonderful. Student built. Great mag… https://t.co/8vRZGBu0zK
RT @laurimnelson: @PublicLab @KyleJwilke The right thing was in the wrong box! This microscope is wonderful. Student built. Great magnifica…
Mon Mar 11 23:03:01 UTC 2019
```

**Comment by jywarren** [comment]
You replied to this and it didn't get copied back, right? https://twitter.com/PublicLab/status/1105873695782957057

**Comment by jywarren** [comment]
I just replied there too, let's see! Shall we log out the text it's searching for a URL in? Maybe it gets cut off?

**Comment by jywarren** [comment]
Lets console out more of this section, and see what it's getting! I think we may need to look specfically for `"https://"+Rails.root` ?

**Comment by namangupta01** [comment]
Yes! I am going to make pull request with `puts` statement at various points.

**Comment by jywarren** [comment]
Hi @namangupta01 -- any updates on this? Thank you!

**Comment by namangupta01** [comment]
Oh Sorry, for missing this out!
Working on it now.
Thanks

**Comment by jywarren** [comment]
Hi @namangupta01 can you take the next step here, we're so close!

**Comment by jywarren** [comment]
OK, so for testing, this tweet has the proper style URL (not converted to `t.co`), so we should get replies from it posted back IF the user has a matching twitter account (which Naman and I do): https://twitter.com/PublicLab/status/1105873695782957057

> Just shared! Check out the map created from our #balloonmapping session at the recent #TXbarnraising in Houston https://publiclab.org/n/18498

**Comment by jywarren** [comment]
Then following that we'd have to check how to deal with the `t.co` converted short URLs, but that'd be a start esp if we manually manage the conversion to shortcodes for now!

**Comment by namangupta01** [comment]
I just tweeted on that tweet. Can you please share the logs so that I can see what can actually be the problem.

**Comment by jywarren** [comment]
OK, i see this:

```
@PublicLab is also talking about @dNOSES_EU from the US - Putting Odor Issues on the Map: Using citizen science to… https://t.co/MMEgTSXUko
RT @Odourcollect: @PublicLab is also talking about @dNOSES_EU from the US - Putting Odor Issues on the Map: Using citizen science to curb o…
RT @Odourcollect: @PublicLab is also talking about @dNOSES_EU from the US - Putting Odor Issues on the Map: Using citizen science to curb o…
@PublicLab Yep. They’ve been sailing around for a few years.
@PublicLab @SenWhitehouse Neat!
RT @Odourcollect: @PublicLab is also talking about @dNOSES_EU from the US - Putting Odor Issues on the Map: Using citizen science to curb o…
@PublicLab Yeah this gif is Amazing
Wed Jul 10 21:31:06 UTC 2019
```

Probably this is yours? `@PublicLab Yeah this gif is Amazing`

**Comment by jywarren** [comment]
So, if you need more info, perhaps console log out more stuff? But be sure not to share anything truly private! Thanks!

**Comment by namangupta01** [comment]
Yup that what I was thinking to do.
Will create pull request soon.

**Comment by jywarren** [comment]
Hi @namangupta01 how is this going? Thank you!

**Comment by namangupta01** [comment]
Hi @jywarren, sorry for missing this out. I have opened the pr lets get that merge. We will be able to check what is happening in there.

**Comment by jywarren** [comment]
awesome, thanks!
